lib/galaxy/util/tool_shed/xml_util.py

```diff
@@ -11,16 +11,25 @@
 log = logging.getLogger(__name__)
 
 
-class Py27CommentedTreeBuilder(XmlET.TreeBuilder):
+class CommentedTreeBuilder(XmlET.TreeBuilder):
+
+    non_comment_seen = False
 
     def doctype(*args):
         # handle deprecation warning for XMLParsing a file with DOCTYPE
         pass
 
+    def start(self, tag, attrib):
+        if tag is not XmlET.Comment:
+            self.non_comment_seen = True
+        super(CommentedTreeBuilder, self).start(tag, attrib)
+
     def comment(self, data):
```

Comment:
When is this method called? It looks like is has been introduced for Python 3.8?
https://docs.python.org/3.8/library/xml.etree.elementtree.html#xml.etree.ElementTree.TreeBuilder.comment

Reply:
I think it's always been called if it existed ? It may just not have been defined previously. If you don't define this you're dropping the comments altogether, which is the point of CommentedTreeBuilder. That's speculation and trying though, I'm not sure how this is implemented.

Reply:
fwiw, the implementation is at https://github.com/python/cpython/blob/master/Lib/xml/etree/ElementTree.py, and we may just use the `insert_comments=True` parameter of the TreeBuilder class on python 3

Reply:
except that then we couldn't use the workaorund of not adding top level comments